solr/core/src/java/org/apache/solr/servlet/PriorityBasedRateLimiter.java

```
@@ -0,0 +1,111 @@
+    super(rateLimiterConfig);
+    this.priorities =
+        new String[] {
+          SolrRequest.RequestPriorities.FOREGROUND.name(),
+          SolrRequest.RequestPriorities.BACKGROUND.name()
+        };
+    this.numRequestsAllowed = new Semaphore(rateLimiterConfig.allowedRequests, true);
+    this.totalAllowedRequests = rateLimiterConfig.allowedRequests;
+    this.waitTimeoutInMillis = rateLimiterConfig.waitForSlotAcquisition;
+  }
+
+  @Override
+  public SlotReservation handleRequest(String requestPriority) {
+    if (!rateLimiterConfig.isEnabled) {
+      return UNLIMITED;
+    }
+    try {
+      if (!acquire(requestPriority)) {
+        return null;
+      }
+    } catch (InterruptedException ie) {
+      return null;
+    }
+    return () -> PriorityBasedRateLimiter.this.release(requestPriority);
+  }
+
+  private boolean acquire(String priority) throws InterruptedException {
+    if (priority.equals(this.priorities[0])) {
+      return nextInQueue();
+    } else if (priority.equals(this.priorities[1])) {
+      if (this.priorityOneRequests.get() < this.totalAllowedRequests) {
+        return nextInQueue();
+      } else {
+        CountDownLatch wait = new CountDownLatch(1);
+        this.waitingList.put(wait);
+        return wait.await(this.waitTimeoutInMillis, TimeUnit.MILLISECONDS) && nextInQueue();
+      }
+    }
+    return true;
+  }
+
+  private boolean nextInQueue() throws InterruptedException {
+    boolean acquired =
+        this.numRequestsAllowed.tryAcquire(1, this.waitTimeoutInMillis, TimeUnit.MILLISECONDS);
+    if (!acquired) {
+      return false;
+    }
+    this.priorityOneRequests.addAndGet(1);
+    return true;
+  }
+
+  private void exitFromQueue() {
+    this.numRequestsAllowed.release(1);
+    this.priorityOneRequests.addAndGet(-1);
+  }
+
+  private void release(String priority) {
+    if (this.priorities[0].equals(priority) || this.priorities[1].equals(priority)) {
```

This is odd -- would it be possible to enforce valid `priority` in `handleRequest`? Then there'd be no need to re-check valid priority here.

As currently written I see that this allows arbitrary other priorities (though notably not `null` -- would produce NPE) to bypass the slot acquisition and just allows the request to proceed. Is there a purpose to doing this way? If a user gets here, shouldn't sending some other priority string result in 4xx client error?

---

Removed in release code. I think now we should throw the exception